agent/wsclient/client.go

```diff
@@ -230,15 +225,51 @@ func (cs *ClientServerImpl) IsReady() bool {
 
 // SetConnection passes a websocket connection object into the client. This is used only in
 // testing and should be avoided in non-test code.
-func (cs *ClientServerImpl) SetConnection(conn WebsocketConn) {
+func (cs *ClientServerImpl) SetConnection(conn wsconn.WebsocketConn) {
 	cs.conn = conn
 }
 
 // SetReadDeadline sets the read deadline for the websocket connection
 // A read timeout results in an io error if there are any outstanding reads
 // that exceed the deadline
 func (cs *ClientServerImpl) SetReadDeadline(t time.Time) error {
-	return cs.conn.SetReadDeadline(t)
+	err := cs.conn.SetReadDeadline(t)
+	if err == nil {
+		return err
+	}
+	seelog.Warnf("Unable to set read deadline for websocket connection: %v for %s", err, cs.URL)
+	// If we get connection closed error from SetReadDeadline, break out of the for loop and
+	// return an error
+	if opErr, ok := err.(*net.OpError); ok && strings.Contains(opErr.Err.Error(), errClosed) {
+		seelog.Errorf("Stopping redundant reads on closed network connection: %s", cs.URL)
+		return opErr
+	}
+	// An unhandled error has occurred while trying to extend read deadline.
```

Comment:
Two things stick out to me here. They're stylistic things, so FFTI:
* This doesn't follow the standard `err != nil` pattern where only good things happen on the left side.
* In my opinion, all of the second half of this should be in a separate method from `SetReadDeadline`. 

Lets say you extracted the server close logic into `closeCSAsync`. I'd maybe go with a structure like this:
```
err := cs.Conn.SetReadDeadline()
if err != nil {
   // check for opErr
       // return opErr
   servErr := closeCSAsync()
   // check for servErr
}
```